pkg/build/admission/defaults/admission.go

```diff
@@ -1,100 +1,91 @@
 package defaults
 
 import (
-	"io"
-
 	"github.com/golang/glog"
-	"k8s.io/kubernetes/pkg/admission"
 	kapi "k8s.io/kubernetes/pkg/api"
-	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
 
 	buildadmission "github.com/openshift/origin/pkg/build/admission"
 	defaultsapi "github.com/openshift/origin/pkg/build/admission/defaults/api"
 	"github.com/openshift/origin/pkg/build/admission/defaults/api/validation"
 	buildapi "github.com/openshift/origin/pkg/build/api"
+	configapi "github.com/openshift/origin/pkg/cmd/server/api"
 )
 
-func init() {
-	admission.RegisterPlugin("BuildDefaults", func(c clientset.Interface, config io.Reader) (admission.Interface, error) {
-
-		defaultsConfig, err := getConfig(config)
-		if err != nil {
-			return nil, err
-		}
-
-		glog.V(5).Infof("Initializing BuildDefaults plugin with config: %#v", defaultsConfig)
-		return NewBuildDefaults(defaultsConfig), nil
-	})
+type BuildDefaults struct {
+	config *defaultsapi.BuildDefaultsConfig
 }
 
-func getConfig(in io.Reader) (*defaultsapi.BuildDefaultsConfig, error) {
-	defaultsConfig := &defaultsapi.BuildDefaultsConfig{}
-	err := buildadmission.ReadPluginConfig(in, defaultsConfig)
+// NewBuildDefaults creates a new BuildDefaults that will apply the defaults specified in the plugin config
+func NewBuildDefaults(pluginConfig map[string]configapi.AdmissionPluginConfig) (BuildDefaults, error) {
+	config := &defaultsapi.BuildDefaultsConfig{}
+	err := buildadmission.ReadPluginConfig(pluginConfig, defaultsapi.BuildDefaultsPlugin, config)
 	if err != nil {
-		return nil, err
+		return BuildDefaults{}, err
 	}
-	errs := validation.ValidateBuildDefaultsConfig(defaultsConfig)
+	errs := validation.ValidateBuildDefaultsConfig(config)
 	if len(errs) > 0 {
-		return nil, errs.ToAggregate()
+		return BuildDefaults{}, errs.ToAggregate()
 	}
-	return defaultsConfig, nil
+	glog.V(4).Infof("Initialized build defaults plugin with config: %#v", *config)
+	return BuildDefaults{config: config}, nil
 }
 
-type buildDefaults struct {
-	*admission.Handler
-	defaultsConfig *defaultsapi.BuildDefaultsConfig
-}
-
-// NewBuildDefaults returns an admission control for builds that sets build defaults
-// based on the plugin configuration
-func NewBuildDefaults(defaultsConfig *defaultsapi.BuildDefaultsConfig) admission.Interface {
-	return &buildDefaults{
-		Handler:        admission.NewHandler(admission.Create),
-		defaultsConfig: defaultsConfig,
-	}
-}
-
-// Admit applies configured build defaults to a pod that is identified
-// as a build pod.
-func (a *buildDefaults) Admit(attributes admission.Attributes) error {
-	if a.defaultsConfig == nil {
-		return nil
-	}
-	if !buildadmission.IsBuildPod(attributes) {
+// ApplyDefaults applies configured build defaults to a build pod
+func (b BuildDefaults) ApplyDefaults(pod *kapi.Pod) error {
```

Comment:
can we take the build as an arg instead of parsing from the annotation? looks like all callers have the build handy (same question applies to ApplyOverrides, I guess)